File: 138. Copy List with Random Pointer/main.py
import copy
from collections import defaultdict
from typing import Optional


class Node:
    def __init__(self, x: int, next: "Node" = None, random: "Node" = None):
        self.val = int(x)
        self.next = next
        self.random = random


"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:
    def copyRandomList(self, head: "Optional[Node]") -> "Optional[Node]":
        if not head:
            return None

        array = list()
        current = head
        count = 0

        while current:
            array.append(Node(current.val))
            count += 1
            current = current.next

        current = head
        index = 0

        while current:
            if current.next:
                array[index].next = array[index + 1]
            if current.random:
                idx = 0
                node = head
                while current.random != node:
                    node = node.next
                    idx += 1
                array[index].random = array[idx]

            current = current.next
            index += 1

        return array[0]


"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:
    def copyRandomList(self, head: "Optional[Node]") -> "Optional[Node]":
        if not head:
            return None

        maps = defaultdict(Optional[Node])

        current = head

        while current:
            maps[current] = Node(current.val)
            current = current.next

        current = head

        while current:
            if current.next:
                maps[current].next = maps[current.next]
            if current.random:
                maps[current].random = maps[current.random]
            current = current.next

        return maps[head]


"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:
    def copyRandomList(self, head: "Optional[Node]") -> "Optional[Node]":
        if not head:
            return None

        maps = defaultdict(Optional[Node])
        maps[None] = None
        current = head

        while current:
            maps[current] = Node(current.val)
            current = current.next

        current = head

        while current:
            maps[current].next = maps[current.next]
            maps[current].random = maps[current.random]
            current = current.next

        return maps[head]


"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:
    def copyRandomList(self, head: "Optional[Node]") -> "Optional[Node]":
        return copy.deepcopy(head)


"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:
    def copyRandomList(self, head: "Optional[Node]") -> "Optional[Node]":
        current = head
        maps = defaultdict(Optional[Node])
        maps[None] = None

        while current:
            if current not in maps:
                maps[current] = Node(current.val)

            if current.next not in maps:
                maps[current.next] = Node(current.next.val)
            maps[current].next = maps[current.next]

            if current.random not in maps:
                maps[current.random] = Node(current.random.val)
            maps[current].random = maps[current.random]

            current = current.next

        return maps[head]

File: 138. Copy List with Random Pointer/main.c
#include <stdio.h>

struct Node {
    int val;
    struct Node *next;
    struct Node *random;
};

#define ARRAY_SIZE 1001

/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *next;
 *     struct Node *random;
 * };
 */
struct Node *copyRandomList(struct Node *head) {
    if (head == NULL) {
        return NULL;
    }

    struct Node **array = (struct Node **)calloc(ARRAY_SIZE, sizeof(struct Node *));
    struct Node *node = head;
    size_t count = 0;

    while (node != NULL) {
        array[count] = (struct Node *)malloc(sizeof(struct Node));
        array[count]->val = node->val;
        array[count]->random = NULL;
        count++;
        node = node->next;
    }
    array[count - 1]->next = NULL;

    node = head;
    size_t current = 0;
    while (node != NULL) {
        array[current]->next = array[current + 1];
        if (node->random != NULL) {
            size_t index = 0;
            struct Node *p = head;
            while (node->random != p) {
                p = p->next;
                index++;
            }
            array[current]->random = array[index];
        }

        current++;
        node = node->next;
    }

    return array[0];
}
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *next;
 *     struct Node *random;
 * };
 */
struct Node *copyRandomList(struct Node *head) {
    if (head == NULL) {
        return NULL;
    }

    struct Node *node = head;

    while (node != NULL) {
        struct Node *copy = (struct Node *)malloc(sizeof(struct Node));
        copy->val = node->val;
        copy->next = node->next;
        copy->random = NULL;
        node->next = copy;
        node = copy->next;
    }

    node = head;

    while (node != NULL) {
        if (node->random != NULL) {
            node->next->random = node->random->next;
        }
        node = node->next->next;
    }

    node = head;
    struct Node *newHead = head->next;

    while (node != NULL) {
        struct Node *copy = node->next;
        node->next = copy->next;
        if (copy->next != NULL) {
            copy->next = copy->next->next;
        }
        node = node->next;
    }

    return newHead;
}


/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *next;
 *     struct Node *random;
 * };
 */
struct Node *copyRandomList(struct Node *head) {
    if (head == NULL) {
        return NULL;
    }

    struct Node *node = head;

    while (node != NULL) {
        struct Node *copy = (struct Node *)malloc(sizeof(struct Node));
        copy->val = node->val;
        copy->next = node->next;
        copy->random = NULL;
        node->next = copy;
        node = copy->next;
    }

    node = head;

    while (node != NULL) {
        if (node->random != NULL) {
            node->next->random = node->random->next;
        }
        node = node->next->next;
    }

    struct Node **current = &head;

    while (*current != NULL) {
        *current = (*current)->next;
        current = &(*current)->next;
    }

    return head;
}